File: core/src/main/javascript/entityauth/EntityAuthenticationData.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>The entity authentication data provides proof of entity identity.</p>
 *
 * <p>Specific entity authentication mechanisms should define their own entity
 * authentication data types.</p>
 *
 * <p>Entity authentication data is represented as
 * {@code
 * entityauthdata = {
 *   "#mandatory" : [ "scheme", "authdata" ],
 *   "scheme" : "string",
 *   "authdata" : object
 * }} where:
 * <ul>
 * <li>{@code scheme} is the entity authentication scheme</li>
 * <li>{@code authdata} is the scheme-specific entity authentication data</li>
 * </ul></p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
	"use strict";
	
	var MslEncodable = require('../io/MslEncodable.js');
	var AsyncExecutor = require('../util/AsyncExecutor.js');
	var MslEntityAuthException = require('../MslEntityAuthException.js');
	var MslError = require('../MslError.js');
	var MslEncoderException = require('../io/MslEncoderException.js');
	var MslEncodingException = require('../MslEncodingException.js');
	
    /** Key entity authentication scheme. */
    var KEY_SCHEME = "scheme";
    /** Key entity authentication data. */
    var KEY_AUTHDATA = "authdata";

    var EntityAuthenticationData = module.exports = MslEncodable.extend({
        /**
         * <p>Create a new entity authentication data object with the specified
         * entity authentication scheme.</p>
         *
         * @param {EntityAuthenticationScheme} scheme the entity authentication
         *        scheme.
         * @constructor
         * @interface
         */
        init: function init(scheme) {
            // The properties.
            var props = {
                scheme: { value: scheme, writable: false, configurable: false },
                /**
                 * Cached encodings.
                 * @type {Object.<MslEncoderFormat,Uint8Array>}
                 */
                encodings: { value: {}, writable: false, enumerable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },

        /**
         * @return {string} the entity identity. May be {@code null} if unknown.
         * @throws MslCryptoException if there is a crypto error accessing the
         *         entity identity.
         */
        getIdentity: function() {},

        /**
         * @param {MslEncoderFactory} encoder MSL encoder factory.
         * @param {MslEncoderFormat} format MSL encoder format.
         * @param {{result: function(MslObject), error: function(Error)}}
         *        callback the callback that will receive the authentication
         *        data MSL representation or any thrown exceptions.
         * @throws MslEncoderException if there was an error constructing the
         *         MSL object.
         */
        getAuthData: function(encoder, format, callback) {},

        /**
         * @param {Object} that the object with which to compare.
         * @return {boolean} true if this object is equal to that object.
         */
        equals: function equals(that) {
            if (this === that) return true;
            if (!(that instanceof EntityAuthenticationData)) return false;
            return this.scheme == that.scheme;
        },

        /** @inheritDoc */
        toMslEncoding: function toMslEncoding(encoder, format, callback) {
        	var self = this;
            AsyncExecutor(callback, function() {
                // Return any cached encoding.
                if (this.encodings[format])
                    return this.encodings[format];
                
                // Get the authentication data.
                this.getAuthData(encoder, format, {
                    result: function(authdata) {
                        AsyncExecutor(callback, function() {
                            // Encode the entity authentication data.
                            var mo = encoder.createObject();
                            mo.put(KEY_SCHEME, this.scheme.name);
                            mo.put(KEY_AUTHDATA, authdata);
                            encoder.encodeObject(mo, format, {
                            	result: function(encoding) {
                            		AsyncExecutor(callback, function() {
                            			// Cache and return the encoding.
                            			this.encodings[format] = encoding;
                            			return encoding;
                            		}, self);
                            	},
                            	error: callback.error,
                            });
                        }, self);
                    },
                    error: callback.error,
                });
            }, self);
        },
    });

    /**
     * Construct a new entity authentication data instance of the correct type
     * from the provided MSL object.
     *
     * @param {MslContext} ctx MSL context.
     * @param {MslObject} entityAuthMo the MSL object.
     * @param {{result: function(EntityAuthenticationData), error: function(Error)}}
     *        callback the callback that will receive the entity authentication
     *        data or any thrown exceptions.
     * @throws MslEntityAuthException if unable to create the entity
     *         authentication data.
     * @throws MslEncodingException if there is an error parsing the entity
     *         authentication data.
     * @throws MslCryptoException if there is an error creating the entity
     *         authentication data crypto.
     */
    var EntityAuthenticationData$parse = function EntityAuthenticationData$parse(ctx, entityAuthMo, callback) {
        AsyncExecutor(callback, function() {
            try {
                var schemeName = entityAuthMo.getString(KEY_SCHEME);
                var encoder = ctx.getMslEncoderFactory();
                var authdata = entityAuthMo.getMslObject(KEY_AUTHDATA, encoder);
        
                // Verify entity authentication scheme.
                var scheme = ctx.getEntityAuthenticationScheme(schemeName);
                if (!scheme)
                    throw new MslEntityAuthException(MslError.UNIDENTIFIED_ENTITYAUTH_SCHEME, schemeName);
        
                // Construct an instance of the concrete subclass.
                var factory = ctx.getEntityAuthenticationFactory(scheme);
                if (!factory)
                    throw new MslEntityAuthException(MslError.ENTITYAUTH_FACTORY_NOT_FOUND, scheme.name);
                factory.createData(ctx, authdata, callback);
            } catch (e) {
                if (e instanceof MslEncoderException)
                    throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "entityauthdata " + entityAuthMo, e);
                throw e;
            }
        });
    };
    
    // Exports.
    module.exports.parse = EntityAuthenticationData$parse;
})(require, (typeof module !== 'undefined') ? module : mkmodule('EntityAuthenticationData'));

File: core/src/main/cpp/entityauth/EntityAuthenticationData.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <entityauth/EntityAuthenticationData.h>
#include <entityauth/EntityAuthenticationFactory.h>
#include <entityauth/EntityAuthenticationScheme.h>
#include <MslEncodingException.h>
#include <MslEntityAuthException.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <io/MslObject.h>
#include <util/MslContext.h>
#include <util/MslUtils.h>
#include <memory>
#include <sstream>

using namespace std;
using namespace netflix::msl;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace entityauth {

namespace {
/** Key entity authentication scheme. */
const string KEY_SCHEME = "scheme";
/** Key entity authentication data. */
const string KEY_AUTHDATA = "authdata";
}

//static
shared_ptr<EntityAuthenticationData> EntityAuthenticationData::create(
        shared_ptr<MslContext> ctx, shared_ptr<MslObject> entityAuthMo)
{
    try {
        // Identify the concrete subclass from the authentication scheme.
        const string schemeName = entityAuthMo->getString(KEY_SCHEME);
        const EntityAuthenticationScheme scheme =
                ctx->getEntityAuthenticationScheme(schemeName);
        if (scheme == EntityAuthenticationScheme::INVALID)
            throw MslEntityAuthException(MslError::UNIDENTIFIED_ENTITYAUTH_SCHEME, schemeName);
        shared_ptr<MslEncoderFactory> encoder = ctx->getMslEncoderFactory();
        shared_ptr<MslObject> authdata = entityAuthMo->getMslObject(KEY_AUTHDATA, encoder);

        // Construct an instance of the concrete subclass.
        shared_ptr<EntityAuthenticationFactory> factory = ctx->getEntityAuthenticationFactory(scheme);
        if (!factory)
            throw MslEntityAuthException(MslError::ENTITYAUTH_FACTORY_NOT_FOUND, scheme.name());
        return factory->createData(ctx, authdata);
    } catch (const MslEncoderException& e) {
        stringstream ss;
        ss << "entityauthdata " << entityAuthMo;
        throw MslEncodingException(MslError::MSL_PARSE_ERROR, ss.str(), e);
    }
}

shared_ptr<ByteArray> EntityAuthenticationData::toMslEncoding(
        shared_ptr<MslEncoderFactory> encoder,
        const MslEncoderFormat& format) const
{
    // Return any cached encoding.
    const MapType::const_iterator it = encodings_.find(format);
    if (it != encodings_.end())
        return it->second;

    // Encode the entity authentication data.
    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put<string>(KEY_SCHEME, scheme_.name());
    mo->put(KEY_AUTHDATA, getAuthData(encoder, format));
    shared_ptr<ByteArray> encoding = encoder->encodeObject(mo, format);

    // Cache and return the encoding.
    encodings_.insert(make_pair(format, encoding));
    return encoding;
}

bool EntityAuthenticationData::equals(shared_ptr<const EntityAuthenticationData> obj) const
{
    if (obj.get() == this) return true;
    return scheme_ == obj->scheme_;
}

bool operator==(const EntityAuthenticationData& a, const EntityAuthenticationData& b)
{
	shared_ptr<const EntityAuthenticationData> ap(&a, &MslUtils::nullDeleter<EntityAuthenticationData>);
	shared_ptr<const EntityAuthenticationData> bp(&b, &MslUtils::nullDeleter<EntityAuthenticationData>);
	return ap->equals(bp);
}

}}} // namespace netflix::msl::entityauth
